src/common/fs/path_util.h

```diff
@@ -0,0 +1,288 @@
+// Copyright 2020 yuzu Emulator Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <filesystem>
+#include <vector>
+
+#include "common/fs/fs_util.h"
+
+namespace Common::FS {
+
+enum class YuzuPath {
+    /// Where yuzu stores its data.
+    YuzuDir,
+    /// Where cached filesystem data is stored.
+    CacheDir,
+    /// Where config files are stored.
+    ConfigDir,
+    /// Where dumped data is stored.
+    DumpDir,
+    /// Where key files are stored.
+    KeysDir,
+    /// Where cheat/mod files are stored.
+    LoadDir,
+    /// Where log files are stored.
+    LogDir,
+    /// Where the emulated NAND is stored.
+    NANDDir,
+    /// Where yuzu screenshots are stored.
+    ScreenshotsDir,
+    /// Where the emulated SDMC is stored.
+    SDMCDir,
+    /// Where shaders are stored.
+    ShaderDir,
+};
+
+/**
+ * Converts a filesystem path to a UTF-8 encoded std::string.
+ *
+ * @param path Filesystem path
+ *
+ * @returns UTF-8 encoded std::string.
+ */
+[[nodiscard]] std::string PathToUTF8String(const std::filesystem::path& path);
+
+/**
+ * Concatenates two filesystem paths together.
+ *
+ * This is needed since the following occurs when using std::filesystem::path's operator/:
+ * first: "/first/path"
+ * second: "/second/path" (Note that the second path has a directory separator in the front)
+ * first / second yields "/second/path" when the desired result is first/path/second/path
+ *
+ * @param first First filesystem path
+ * @param second Second filesystem path
+ *
+ * @returns A concatenated filesystem path.
+ */
+[[nodiscard]] std::filesystem::path ConcatPath(const std::filesystem::path& first,
+                                               const std::filesystem::path& second);
+
+template <typename Path1, typename Path2>
+[[nodiscard]] std::filesystem::path ConcatPath(Path1 first, Path2 second) {
```

Comment:
```suggestion
[[nodiscard]] std::filesystem::path ConcatPath(const Path1& first, const Path2& second) {
```

Ditto elswhere this is applicable